packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js

```diff
@@ -144,38 +145,28 @@ const generateFileRecord = ({ file, granuleCumulusId }) => ({
  *   Cumulus ID of the granule for this file
  * @returns {Array<Object>} - file records
  */
-const generateFileRecords = async ({
+const _generateFileRecords = async ({
   files,
   granuleCumulusId,
-}) => {
-  const errors = [];
-
-  const generatedFiles = files.map((file) => {
-    let generatedFile;
+}) => files.map((file) => generateFileRecord({ file, granuleCumulusId }));
 
-    try {
-      generatedFile = generateFileRecord({ file, granuleCumulusId });
-    } catch (error) {
-      errors.push(error);
-    }
-
-    return generatedFile;
-  });
-
-  return {
-    files: generatedFiles,
-    errors: errors,
-  };
-};
-
-const writeFilesViaTransaction = async ({
+/**
+ * Write an array of file records to the database
+ *
+ * @param {Object} params
+ * @param {Object} params.fileRecords - File objects
+ * @param {Knex} params.knex - Client to interact with Postgres database
+ * @param {Object} params.filePgModel - Optional File model override
+ * @returns {Promise} - Promise resolved once all file upserts resolve
+ */
+const _writeFiles = async ({
   fileRecords,
-  trx,
+  knex,
   filePgModel = new FilePgModel(),
-}) =>
-  Promise.all(fileRecords.map(
-    (fileRecord) => filePgModel.upsert(trx, fileRecord)
-  ));
+}) => pMap(
+  fileRecords,
+  async (fileRecord) => filePgModel.upsert(knex, fileRecord)
```

Comment:
Note we're using `knex` instead of `trx` here. This is based on our finding that a Knex transaction always fails whenever any part of it fails, even if we catch the error.

I updated the tests to try a granule with 10 files (and adhoc tested with 100). It performs fine but it will get slower once you hit 100+ files. Want to flag this in case it seems like something worth addressing now.

Reply:
Hmm, yeah. So the problem is that if we used `trx` here, then as soon as one of the file writes failed, the others would fail. 

Is that a problem though? It's okay if only some of the file writes succeed. I think the main thing is that we want the granule to update to `failed` even if some/all of the file writes fail. Meaning that the granule write has to use `knex` and not `trx`. But maybe we can use `trx` for the file writes.

I'm mentioning it because I think using `knex` vs `trx` can result in a proliferation of database connections, which if you're running thousands of workflows to ingest data (very realistic), we could eat up the number of allowed connections

Reply:
Right. The question here is: would we be ok failing ALL file writes for a granule if ANY file write fails? 

The logic we'd talked about (which I think makes a lot of sense) is basically storing any files that we can, even if some others fail. There doesn't seem to be a way to do that without making each file it's own independent transaction and I totally agree that that seems scary. I've verified that it _works_ as-is for 100+ files but I think, at the very least, once you get into the thousands you'd see a performance hit compared to transactions (unverified but seems really probable).

So from the implementation side it does seem like a better idea to use a transaction but that's different business logic.

Thoughts? Also looping in @Jkovarik because this is sort of a compromise if we change it.

Reply:
I think we've landed on an agreement here. We will keep this as-is where each file write uses the Knex client directly, rather than a transaction. In testing it appears to not cause a massive influx of database connections and it gets the file integrity we're after.